### ISSUE TITLE
Zero member tuple struct should warn on incorrect construction

### ISSUE THREAD

**Issue by edward-shen** [rust-lang/rust#87181]
Hey folks, this just bit me so and took a bit to figure out, so I'd thought I'd file an issue for this admittedly niche issue.

Given the following code: [playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=b709d444dfcf0590f293531e3a7e173f)

```rust
trait Foo {
    fn get(&self);
}

struct Bar<T>(T);

struct Inner();

impl Foo for Inner {
    fn get(&self) { }
}

fn main() {
    let thing = Bar(Inner);
    thing.0.get();
}
```

The current output is:

```
error[E0599]: no method named `get` found for fn item `fn() -> Inner {Inner}` in the current scope
  --> src/main.rs:16:13
   |
16 |     thing.0.get();
   |             ^^^ method not found in `fn() -> Inner {Inner}`
   |
   = note: `thing.0` is a function, perhaps you wish to call it
   = help: items from traits can only be used if the trait is implemented and in scope
note: `Foo` defines an item `get`, perhaps you need to implement it
  --> src/main.rs:2:1
   |
2  | trait Foo {
   | ^^^^^^^^^

error: aborting due to previous error

For more information about this error, try `rustc --explain E0599`.
error: could not compile `playground`

To learn more, run the command again with --verbose.
```

Ideally the output should look like:

```
error[E0599]: no method named `get` found for fn item `fn() -> Inner {Inner}` in the current scope
  --> src/main.rs:16:13
   |
16 |     thing.0.get();
   |             ^^^ method not found in `fn() -> Inner {Inner}`
   |
   = note: `thing.0` is a function, perhaps you wish to call it
   = note: `Inner` is a tuple struct with zero fields, perhaps you wish to initialize it instead: `Bar(Inner())`
   = help: items from traits can only be used if the trait is implemented and in scope
note: `Foo` defines an item `get`, perhaps you need to implement it
  --> src/main.rs:2:1
   |
2  | trait Foo {
   | ^^^^^^^^^

error: aborting due to previous error

For more information about this error, try `rustc --explain E0599`.
error: could not compile `playground`

To learn more, run the command again with --verbose.
```

This was an extremely confusing error message for me, as I suddenly had a `fn() -> Inner` type out of no where in my code! Of course, this was a simplified issue, but it's not immediately obvious that 1. a unit struct is actually a function returning its own type, and 2. the problem was due to the fact that a unit struct is not the same as a zero-member tuple struct.

This was extracted out of a much more complex error message, where the only hint to the actual problem was that the unit struct "constructor" didn't satisfy the trait bound, which pointed to a construction issue.

Not quite sure whether or not it's possible to actually implement a suggestion as written, but perhaps even a check between `Inner` vs `Inner()` might be a good note to have?

**Comment by estebank** [comment]
We should [suggest writing `(self.0)()`](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=fac439b4d516b3f36ad4302e6c40087d) like we already do if you write `self.0()`.

You'll want to look at this file

https://github.com/rust-lang/rust/blob/2119976c492894b72287f08865c71d63cff8d471/compiler/rustc_typeck/src/astconv/mod.rs#L1714-L1751

and a check to see if `qself_ty` is a closure. I believe that `hir_ref_id` is the node for `thing.0.get()`, but I am not certain. If it is, then you can use `tcx.hir().find(hir_ref_id)` to get the `hir::Node`, from which you can decompose it to get the span for `thing.0`, and use it for a `multipart_suggestion` to suggest the addition of parentheses.

**Comment by hi-rustin** [comment]
I may not have the time these days, maybe others would like to try and unassign myself.

**Comment by JaydenElliott** [comment]
Happy to try give this a go as my first contribution.

**Comment by estebank** [comment]
@JaydenElliott Yes! Feel free to ping me on zulip.

**Comment by inashivb** [comment]
Hi @JaydenElliott !
Are you still working on this issue? (I am looking for issues to work on so just checking in :) )

**Comment by JaydenElliott** [comment]
Hi @inashivb, I think this one was a bit too difficult for me, so feel free to take it off of me!

**Comment by inashivb** [comment]
Thank you very much, @JaydenElliott ! I'm not sure if I'll be able to do it either but I'd like to give it a try. I hope you find a new issue to work on soon. :)

**Comment by George-lewis** [comment]
Hello everyone, this issue seems to have stalled, so I think I'll take a crack at it myself if that's okay

**Comment by George-lewis** [comment]
It took me a lot of digging, but I've landed in `rustc_typeck::check::method::suggest::report_method_error`. This function is called when there was an error around a method, including when it's not found. So, in the case where the method isn't found, I check if the receiver is a no-argument constructor and apply the suggestion. I'm still working on it, but I think I'll open a PR soon-ish

**Comment by George-lewis** [comment]
I've found a related issue for field access on an unconstructed struct. It can be seen here:

```rs
struct Bar<T>(T);

struct Inner(u8);

fn main() {
    let thing = Bar(Inner);
    thing.0.0();
}
```

Which yields: `error[E0609]: no field `0` on type `fn(u8) -> Inner {Inner}``

We could probably do something similar to what we're doing here, and tell them to construct the struct/enum variant? Does anyone know if there's an issue for this, and can I fix it here?
